papers/oleksandr_pavlyk/optimizations.rst

```diff
@@ -72,9 +72,129 @@ The most significant improvements in memory optimizations in Update 2 comes from
+        args = argParser.parse_args()
+
+        REP = args.iteration 
+
+        try:
+            from daal.services import Environment
+            nThreadsInit = Environment.getInstance().getNumberOfThreads()
+            Environment.getInstance().setNumberOfThreads(args.proc)
+        except:
+            pass
+
+        def st_time(func):
+            def st_func(*args, **keyArgs):
+                times = []
+                for n in range(REP):
+                    t1 = timeit.default_timer()
+                    r = func(*args, **keyArgs)
+                    t2 = timeit.default_timer()
+                    times.append(t2-t1)
+                print (min(times), end='')
+                return r
+            return st_func
+
+        problem_sizes = [
+                (10000,   2),
+                (10000,   25),
+                (10000,   50),
+                (50000,   2),
+                (50000,   25),
+                (50000,   50),
+                (100000,  2),
+                (100000,  25),
+                (100000,  50)]
+
+        X={}
+        for p, n in problem_sizes:
+            X[(p,n)] = rand(p,n)
+
+
+        kmeans = KMeans(n_clusters=10, n_jobs=args.proc)
+        @st_time
+        def train(X):
+            kmeans.fit(X)
+
+        for p, n in problem_sizes:
+            print (p,n, end=' ')
+            X_local = X[(p,n)]
+            train(X_local)
+            print('')
+
+Using all 32 cores of Xeon E5-2698v3 IDP’s KMeans can be faster more than 50 times comparing with python available on Ubuntu 14.04.
+P below means amount of CPU cores used.
+
+.. table:: 
+
+ +------+----+---------+----------+------------+-------------+-------------+--------------+
+ |rows  |cols|IDP,s P=1|IDP,s P=32|System,s P=1|System,s P=32|Vs System,P=1|Vs System,P=32|
+ +======+====+=========+==========+============+=============+=============+==============+
+ |10000 |2   |0,01     |0,01      |0,38        |0,27         |28,55        |36,52         |
+ |10000 |25  |0,05     |0,01      |1,46        |0,57         |27,59        |48,22         |
+ |10000 |50  |0,09     |0,02      |2,21        |0,87         |23,83        |40,76         |
+ |50000 |2   |0,08     |0,01      |1,62        |0,57         |20,57        |47,43         |
+ |50000 |25  |0,67     |0,07      |14,43       |2,79         |21,47        |38,69         |
+ |50000 |50  |1,05     |0,10      |24,04       |4,00         |22,89        |38,52         |
+ |100000|2   |0,15     |0,02      |3,33        |0,87         |22,30        |56,72         |
+ |100000|25  |1,34     |0,11      |33,27       |5,53         |24,75        |49,07         |
+ |100000|50  |2,21     |0,17      |63,30       |8,36         |28,65        |47,95         |
+ +------+----+---------+----------+------------+-------------+-------------+--------------+
+
+We compared the similar runs for other algorithms and normalized results by results obtained with DAAL in C++ without python to estimate overhead from python wrapping.
```

Comment:
"We performed similar runs for other algorithms, reporting execution times as a fraction of execution time of C++ native program that uses DAAL in order to estimate the overhead of using Python wrappers."